templates/default/deploy-scripts/src/deployer.rs

```diff
@@ -0,0 +1,394 @@
+        Energy, TransactionType, WalletAccount,
+    },
+    v2::{self, BlockIdentifier},
+};
+use std::path::Path;
+use std::sync::Arc;
+
+/// A struct containing connection and wallet information.
+#[derive(Debug)]
+pub struct Deployer {
+    /// The client to establish a connection to a Concordium node (V2 API).
+    pub client: v2::Client,
+    /// The account keys to be used for sending transactions.
+    pub key: Arc<WalletAccount>,
+}
+
+impl Deployer {
+    /// A function to create a new deployer instance from a network client and a path to the wallet.
+    pub fn new(client: v2::Client, wallet_account_file: &Path) -> Result<Deployer, Error> {
+        let key_data = WalletAccount::from_json_file(wallet_account_file)
+            .context("Unable to read wallet file.")?;
+
+        Ok(Deployer {
+            client,
+            key: key_data.into(),
+        })
+    }
+
+    /// A function to check if a module exists on the chain.
+    pub async fn module_exists(
+        &mut self,
+        module_reference: &ModuleReference,
+    ) -> Result<bool, Error> {
+        let module_src = self
+            .client
+            .get_module_source(module_reference, &BlockIdentifier::LastFinal)
+            .await;
+
+        match module_src {
+            Ok(_) => Ok(true),
+            Err(e) if e.is_not_found() => Ok(false),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    /// A function to deploy a wasm module on the chain.
+    ///
+    /// If successful, the transaction hash, the block item, and
+    /// the module reference are returned.
+    /// If the module already exists on
+    /// chain, this function returns the module reference of the already
+    /// deployed module (no transaction hash or the block item returned).
+    ///
+    /// An optional expiry time for the transaction
+    /// can be given. If `None` is provided, the local time + 300 seconds is
+    /// used as a default expiry time.
+    pub async fn deploy_wasm_module(
+        &mut self,
+        wasm_module: WasmModule,
+        expiry: Option<TransactionTime>,
+    ) -> Result<
+        (
+            Option<TransactionHash>,
+            Option<BlockItemSummary>,
+            ModuleReference,
+        ),
+        Error,
+    > {
+        println!("\nDeploying module....");
+
+        let module_reference = wasm_module.get_module_ref();
+
+        let exists = self.module_exists(&module_reference).await?;
+
+        if exists {
+            println!(
+                "Module with reference {} already exists on the chain.",
+                module_reference
+            );
+            return Ok((None, None, module_reference));
+        }
+
+        let nonce = self.get_nonce(self.key.address).await?;
+
+        if !nonce.all_final {
+            anyhow::bail!("Nonce not final")
+        }
+
+        let expiry = expiry.unwrap_or(TransactionTime::from_seconds(
+            (chrono::Utc::now().timestamp() + 300) as u64,
+        ));
```

Comment:
Are you sure you ran clippy on this script? I would expect it to complain about this and suggest to use `unwrap_or_else`.

Reply:
There are no warnings by Clippy. Both options are fine and I chose the latter one because it is a bit less code.

```
expiry.unwrap_or_else( || TransactionTime::from_seconds(
            (chrono::Utc::now().timestamp() + 300) as u64,
        )
```
or 
```
expiry.unwrap_or( TransactionTime::from_seconds(
            (chrono::Utc::now().timestamp() + 300) as u64,
        )
```

Reply:
It's not that it's less code, the problem is that in the option `unwrap_or` you are making a call `Utc::now` to get the current time **always**, not just in the case it is not supplied. Both of them are correct in this case though.

So I'm surprised clippy does not complain, it usually does when there's a complex calculation in the unwrap_or clause.